improver/lapse_rate.py

```diff
@@ -383,59 +299,50 @@ def _generate_lapse_rate_array(
         )
 
-        orog_nbhoods = np.zeros(
-            (dataarray_size, self.nbhoodarray_size), dtype=np.float32
-        )
-        fnc = SaveNeighbourhood(allbuffers=orog_nbhoods)
-        generic_filter(
-            orography_data,
-            fnc.filter,
-            size=self.nbhood_size,
-            mode="constant",
-            cval=np.nan,
-        )
+        # Zips together the windows for temperature and orography
+        # then finds the gradient of the surface temperature with
+        # orography height - i.e. lapse rate.
+        axis = (-2, -1)
+        for temp, orog in zip(temp_nbhood_window, orog_nbhood_window):
+            # height_diff is True for points where the height
+            # difference between the central points and its
+            # neighbours is < max_height_diff.
+            height_diff_mask = self._create_height_diff_mask(orog)
+            temp = np.where(height_diff_mask, temp, np.nan)
+
+            # Places NaNs in orog to match temp.
+            orog = np.where(np.isnan(temp), np.nan, orog)
+
+            grad = mathematical_operations.alinfit(
+                orog, temp, axis=axis, gradient_only=True
+            )
 
-        # height_diff_mask is True for points where the height
-        # difference between the central point and its neighbours
-        # is > max_height_diff.
-        height_diff_mask = self._create_heightdiff_mask(orog_nbhoods)
-
-        # Mask points with extreme height differences as NaN.
-        temp_nbhoods = np.where(height_diff_mask, np.nan, temp_nbhoods)
-        orog_nbhoods = np.where(height_diff_mask, np.nan, orog_nbhoods)
-
-        # Loop through both arrays and find gradient of surface temperature
-        # with orography height - ie lapse rate.
-        # TODO: This for loop is the bottleneck in the code and needs to
-        # be parallelised.
-        lapse_rate_array = [
-            self._calc_lapse_rate(temp, orog)
-            for temp, orog in zip(temp_nbhoods, orog_nbhoods)
-        ]
-
-        lapse_rate_array = np.array(lapse_rate_array, dtype=np.float32).reshape(
-            (temperature_data.shape)
-        )
+            # Checks that the standard deviations are not 0
+            # i.e. there is some variance to fit a gradient to.
+            tempcheck = np.isclose(np.nanstd(temp, axis=axis), 0)
+            orogcheck = np.isclose(np.nanstd(orog, axis=axis), 0)
+            grad = np.where(tempcheck | orogcheck, DALR, grad)
+            # checks that our central point in the neighbourhood
+            # is not NaN.
+            temp_nan_check = np.isnan(
+                temp[..., self.ind_central_point, self.ind_central_point]
+            )
+            grad = np.where(temp_nan_check, DALR, grad)
+
+            lapse_rate_array.append(grad)
+
+        lapse_rate_array = np.array(lapse_rate_array, dtype=np.float32)
 
         # Enforce upper and lower limits on lapse rate values.
         lapse_rate_array = np.where(
```

Comment:
Could use ` lapse_rate_array.clip(self.min_lapse_rate, self.max_lapse_rate)` here instead of `np.where`.

Reply:
Nice! That's much cleaner